File: packages/sdk-go/melaya/pipelines.go
// Pipelines API — embeddable lifecycle (build/run/read) plus overview, traces,
// and schedules.
//
// Embeddable lifecycle:  /api/v1/private/pipelines/*
// Overview:             /api/v1/private/overview/pipeline*
// Traces:               /api/v1/private/runs/:runId/traces
// Schedules:            /api/v1/private/pipeline-schedule
package melaya

import (
	"context"
	"fmt"
	"net/url"
	"strings"
)

// PipelinesAPI wraps pipeline embeddable lifecycle, overview, trace, and
// schedule endpoints.
type PipelinesAPI struct {
	h *httpClient
}

// ── Embeddable lifecycle: build → run → read ──────────────────────────────────
//
// Full CRUD + run + outputs over /api/v1/private/pipelines/*. Every method is
// tier-gated server-side (Forge+ to create/run) and tenant-scoped to the
// caller's projects.

// ListPipelines returns all pipelines visible to the caller.
//
// GET /api/v1/private/pipelines
func (p *PipelinesAPI) ListPipelines(ctx context.Context) ([]PipelineConfig, error) {
	data, err := p.h.get(ctx, "/api/v1/private/pipelines", nil)
	if err != nil {
		return nil, err
	}
	var env struct {
		Pipelines []PipelineConfig `json:"pipelines"`
	}
	if err := unmarshal(data, &env); err != nil {
		return nil, err
	}
	return env.Pipelines, nil
}

// Create creates a new pipeline. name and project are required; include the
// full agent config (agents, prompts, models, wiring) in the same body map.
//
// POST /api/v1/private/pipelines
func (p *PipelinesAPI) Create(ctx context.Context, body PipelineConfig) (PipelineConfig, error) {
	data, err := p.h.post(ctx, "/api/v1/private/pipelines", body)
	if err != nil {
		return nil, err
	}
	var v PipelineConfig
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return v, nil
}

// Get returns the full config for one pipeline. project is optional and narrows
// the tenant scope.
//
// GET /api/v1/private/pipelines/{name}?project=
func (p *PipelinesAPI) Get(ctx context.Context, name, project string) (PipelineConfig, error) {
	q := map[string]string{}
	if project != "" {
		q["project"] = project
	}
	path := "/api/v1/private/pipelines/" + url.PathEscape(name)
	data, err := p.h.get(ctx, path, q)
	if err != nil {
		return nil, err
	}
	var v PipelineConfig
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return v, nil
}

// Update replaces a pipeline's config. Pass the full config plus the owning
// project. This is the path for editing per-agent prompts or swapping models.
//
// PUT /api/v1/private/pipelines/{name}
func (p *PipelinesAPI) Update(ctx context.Context, name string, body PipelineUpdateBody) (PipelineConfig, error) {
	path := "/api/v1/private/pipelines/" + url.PathEscape(name)
	data, err := p.h.put(ctx, path, body)
	if err != nil {
		return nil, err
	}
	var v PipelineConfig
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return v, nil
}

// Delete deletes a pipeline.
//
// DELETE /api/v1/private/pipelines/{name}?project=
func (p *PipelinesAPI) Delete(ctx context.Context, name, project string) error {
	q := map[string]string{}
	if project != "" {
		q["project"] = project
	}
	path := "/api/v1/private/pipelines/" + url.PathEscape(name)
	_, err := p.h.del(ctx, path, q)
	return err
}

// Run triggers a pipeline run. Returns immediately with a run_id; subscribe to
// progress via Platform.Events.OnRunUpdate or poll RunStatus.
//
// POST /api/v1/private/pipelines/{name}/run
func (p *PipelinesAPI) Run(ctx context.Context, name string, opts *PipelineRunOptions) (*PipelineRunAccepted, error) {
	var body interface{} = map[string]interface{}{}
	if opts != nil {
		body = opts
	}
	path := "/api/v1/private/pipelines/" + url.PathEscape(name) + "/run"
	data, err := p.h.post(ctx, path, body)
	if err != nil {
		return nil, err
	}
	var v PipelineRunAccepted
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return &v, nil
}

// RunIDs lists run IDs for a pipeline, filtered to the caller's tier retention window.
//
// GET /api/v1/private/pipelines/{name}/runs
func (p *PipelinesAPI) RunIDs(ctx context.Context, name string) ([]string, error) {
	path := "/api/v1/private/pipelines/" + url.PathEscape(name) + "/runs"
	data, err := p.h.get(ctx, path, nil)
	if err != nil {
		return nil, err
	}
	var env struct {
		RunIDs []string `json:"run_ids"`
	}
	if err := unmarshal(data, &env); err != nil {
		return nil, err
	}
	return env.RunIDs, nil
}

// RunStatus returns the status and best-effort cost for one run.
//
// GET /api/v1/private/pipelines/{name}/runs/{runID}
func (p *PipelinesAPI) RunStatus(ctx context.Context, name, runID string) (*PipelineRunStatus, error) {
	path := "/api/v1/private/pipelines/" + url.PathEscape(name) + "/runs/" + url.PathEscape(runID)
	data, err := p.h.get(ctx, path, nil)
	if err != nil {
		return nil, err
	}
	var v PipelineRunStatus
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return &v, nil
}

// CancelRun cancels or kills an in-flight run.
//
// DELETE /api/v1/private/pipelines/{name}/runs/{runID}
func (p *PipelinesAPI) CancelRun(ctx context.Context, name, runID string) error {
	path := "/api/v1/private/pipelines/" + url.PathEscape(name) + "/runs/" + url.PathEscape(runID)
	_, err := p.h.del(ctx, path, nil)
	return err
}

// Outputs lists the artifacts a pipeline has produced.
//
// GET /api/v1/private/pipelines/{name}/outputs
func (p *PipelinesAPI) Outputs(ctx context.Context, name string) (interface{}, error) {
	path := "/api/v1/private/pipelines/" + url.PathEscape(name) + "/outputs"
	data, err := p.h.get(ctx, path, nil)
	if err != nil {
		return nil, err
	}
	var v interface{}
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return v, nil
}

// Output reads one output artifact by relative path. Each path segment is
// url-path-escaped individually. Pass download=true to force a download
// disposition on the wire.
//
// GET /api/v1/private/pipelines/{name}/outputs/{path}
func (p *PipelinesAPI) Output(ctx context.Context, name, artifactPath string, download bool) (interface{}, error) {
	segments := strings.Split(artifactPath, "/")
	escaped := make([]string, len(segments))
	for i, seg := range segments {
		escaped[i] = url.PathEscape(seg)
	}
	path := "/api/v1/private/pipelines/" + url.PathEscape(name) + "/outputs/" + strings.Join(escaped, "/")
	q := map[string]string{}
	if download {
		q["download"] = "1"
	}
	data, err := p.h.get(ctx, path, q)
	if err != nil {
		return nil, err
	}
	var v interface{}
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return v, nil
}

// PreviewCode generates a read-only code preview for a config (no persistence,
// no tier gate).
//
// POST /api/v1/private/pipelines/preview-code
func (p *PipelinesAPI) PreviewCode(ctx context.Context, config PipelineConfig) (interface{}, error) {
	data, err := p.h.post(ctx, "/api/v1/private/pipelines/preview-code", config)
	if err != nil {
		return nil, err
	}
	var v interface{}
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return v, nil
}

// Tools returns the builder tool registry available to pipeline agents.
//
// GET /api/v1/private/pipelines/tools
func (p *PipelinesAPI) Tools(ctx context.Context) (interface{}, error) {
	data, err := p.h.get(ctx, "/api/v1/private/pipelines/tools", nil)
	if err != nil {
		return nil, err
	}
	var v interface{}
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return v, nil
}

// Subagents returns the builder sub-agent/crew registry.
//
// GET /api/v1/private/pipelines/subagents
func (p *PipelinesAPI) Subagents(ctx context.Context) (interface{}, error) {
	data, err := p.h.get(ctx, "/api/v1/private/pipelines/subagents", nil)
	if err != nil {
		return nil, err
	}
	var v interface{}
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return v, nil
}

// InstantiateTemplate instantiates a pipeline from a template the caller can
// access. Merges the template payload with overrides, then creates the pipeline
// under name + project.
//
// POST /api/v1/private/templates/{templateID}/instantiate
func (p *PipelinesAPI) InstantiateTemplate(ctx context.Context, templateID string, body TemplateInstantiateBody) (*TemplateInstantiateResult, error) {
	path := "/api/v1/private/templates/" + url.PathEscape(templateID) + "/instantiate"
	data, err := p.h.post(ctx, path, body)
	if err != nil {
		return nil, err
	}
	var v TemplateInstantiateResult
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return &v, nil
}

// BuildWithAI builds a pipeline from a natural-language brief (synchronous —
// returns the generated config). For streamed progress, use the SSE endpoint
// /api/v1/private/ai/build-pipeline directly.
//
// POST /api/v1/private/ai/build-pipeline/sync
func (p *PipelinesAPI) BuildWithAI(ctx context.Context, brief map[string]interface{}) (interface{}, error) {
	data, err := p.h.post(ctx, "/api/v1/private/ai/build-pipeline/sync", brief)
	if err != nil {
		return nil, err
	}
	var v interface{}
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return v, nil
}

// ── Overview ─────────────────────────────────────────────────────────────────

// Overview returns the dashboard summary (usage stats, active strategies, recent runs).
//
// GET /api/v1/private/overview
func (p *PipelinesAPI) Overview(ctx context.Context) (*OverviewSummary, error) {
	data, err := p.h.get(ctx, "/api/v1/private/overview", nil)
	if err != nil {
		return nil, err
	}
	var v OverviewSummary
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return &v, nil
}

// Count returns pipeline run counts grouped by status.
//
// GET /api/v1/private/overview/pipeline-count
func (p *PipelinesAPI) Count(ctx context.Context) (*PipelineCountByStatus, error) {
	data, err := p.h.get(ctx, "/api/v1/private/overview/pipeline-count", nil)
	if err != nil {
		return nil, err
	}
	var v PipelineCountByStatus
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return &v, nil
}

// List returns a paginated list of pipeline runs.
//
// GET /api/v1/private/overview/pipelines
func (p *PipelinesAPI) List(ctx context.Context, params *PipelineListParams) ([]PipelineRun, error) {
	q := map[string]string{}
	if params != nil {
		if params.Project != "" {
			q["project"] = params.Project
		}
		if params.PipelineName != "" {
			q["pipelineName"] = params.PipelineName
		}
		if params.Status != "" {
			q["status"] = params.Status
		}
		if params.Limit != nil {
			q["limit"] = fmt.Sprintf("%d", *params.Limit)
		}
		if params.Offset != nil {
			q["offset"] = fmt.Sprintf("%d", *params.Offset)
		}
		if params.Page != nil {
			q["page"] = fmt.Sprintf("%d", *params.Page)
		}
	}
	data, err := p.h.get(ctx, "/api/v1/private/overview/pipelines", q)
	if err != nil {
		return nil, err
	}
	var v []PipelineRun
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return v, nil
}

// Recent returns the most recent pipeline runs for dashboard widgets.
//
// GET /api/v1/private/overview/pipelines/recent
func (p *PipelinesAPI) Recent(ctx context.Context) ([]PipelineRun, error) {
	data, err := p.h.get(ctx, "/api/v1/private/overview/pipelines/recent", nil)
	if err != nil {
		return nil, err
	}
	var v []PipelineRun
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return v, nil
}

// ── Traces ───────────────────────────────────────────────────────────────────

// Traces returns the paginated trace list for a run.
// The server returns { data: { list, total, page, pageSize } }.
//
// GET /api/v1/private/runs/:runId/traces
func (p *PipelinesAPI) Traces(ctx context.Context, runID string) (*TracePage, error) {
	path := "/api/v1/private/runs/" + url.PathEscape(runID) + "/traces"
	data, err := p.h.get(ctx, path, nil)
	if err != nil {
		return nil, err
	}
	var env struct {
		Data TracePage `json:"data"`
	}
	if err := unmarshal(data, &env); err != nil {
		return nil, err
	}
	return &env.Data, nil
}

// Trace returns a single trace by ID.
//
// GET /api/v1/private/runs/:runId/traces/:traceId
func (p *PipelinesAPI) Trace(ctx context.Context, runID, traceID string) (*Trace, error) {
	path := "/api/v1/private/runs/" + url.PathEscape(runID) + "/traces/" + url.PathEscape(traceID)
	data, err := p.h.get(ctx, path, nil)
	if err != nil {
		return nil, err
	}
	var v Trace
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return &v, nil
}

// TraceStats returns statistics for a specific trace.
//
// GET /api/v1/private/runs/:runId/traces/:traceId/stats
func (p *PipelinesAPI) TraceStats(ctx context.Context, runID, traceID string) (*TraceStats, error) {
	path := "/api/v1/private/runs/" + url.PathEscape(runID) + "/traces/" + url.PathEscape(traceID) + "/stats"
	data, err := p.h.get(ctx, path, nil)
	if err != nil {
		return nil, err
	}
	var v TraceStats
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return &v, nil
}

// DeleteTraces deletes all traces for a run by runId (no body required).
// Returns the number of spans deleted.
//
// DELETE /api/v1/private/runs/:runId/traces
func (p *PipelinesAPI) DeleteTraces(ctx context.Context, runID string) (*DeleteTracesResult, error) {
	path := "/api/v1/private/runs/" + url.PathEscape(runID) + "/traces"
	data, err := p.h.del(ctx, path, nil)
	if err != nil {
		return nil, err
	}
	var v DeleteTracesResult
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return &v, nil
}

// ── Schedules ────────────────────────────────────────────────────────────────

// ListSchedules lists all pipeline schedules accessible to the caller.
//
// GET /api/v1/private/pipeline-schedule
func (p *PipelinesAPI) ListSchedules(ctx context.Context) ([]PipelineSchedule, error) {
	data, err := p.h.get(ctx, "/api/v1/private/pipeline-schedule", nil)
	if err != nil {
		return nil, err
	}
	var v []PipelineSchedule
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return v, nil
}

// GetSchedule returns the schedule status for a specific pipeline.
//
// GET /api/v1/private/pipeline-schedule/:project/:pipelineName
func (p *PipelinesAPI) GetSchedule(ctx context.Context, project, pipelineName string) (*PipelineSchedule, error) {
	path := "/api/v1/private/pipeline-schedule/" + url.PathEscape(project) + "/" + url.PathEscape(pipelineName)
	data, err := p.h.get(ctx, path, nil)
	if err != nil {
		return nil, err
	}
	var v PipelineSchedule
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return &v, nil
}

// UpsertSchedule creates or updates a pipeline schedule.
//
// PUT /api/v1/private/pipeline-schedule/:project/:pipelineName
func (p *PipelinesAPI) UpsertSchedule(ctx context.Context, project, pipelineName string, body PipelineScheduleUpsertBody) (*PipelineSchedule, error) {
	path := "/api/v1/private/pipeline-schedule/" + url.PathEscape(project) + "/" + url.PathEscape(pipelineName)
	data, err := p.h.put(ctx, path, body)
	if err != nil {
		return nil, err
	}
	var v PipelineSchedule
	if err := unmarshal(data, &v); err != nil {
		return nil, err
	}
	return &v, nil
}

// PauseSchedule pauses a pipeline schedule.
//
// POST /api/v1/private/pipeline-schedule/:project/:pipelineName/pause
func (p *PipelinesAPI) PauseSchedule(ctx context.Context, project, pipelineName string) (bool, error) {
	path := "/api/v1/private/pipeline-schedule/" + url.PathEscape(project) + "/" + url.PathEscape(pipelineName) + "/pause"
	data, err := p.h.post(ctx, path, nil)
	if err != nil {
		return false, err
	}
	var v okResult
	if err := unmarshal(data, &v); err != nil {
		return false, err
	}
	return v.Ok, nil
}

// ResumeSchedule resumes a paused pipeline schedule.
//
// POST /api/v1/private/pipeline-schedule/:project/:pipelineName/resume
func (p *PipelinesAPI) ResumeSchedule(ctx context.Context, project, pipelineName string) (bool, error) {
	path := "/api/v1/private/pipeline-schedule/" + url.PathEscape(project) + "/" + url.PathEscape(pipelineName) + "/resume"
	data, err := p.h.post(ctx, path, nil)
	if err != nil {
		return false, err
	}
	var v okResult
	if err := unmarshal(data, &v); err != nil {
		return false, err
	}
	return v.Ok, nil
}

File: packages/sdk/src/pipelines.ts
/**
 * Pipelines API — overview, runs, traces, and schedules.
 *
 * Covers the `/api/v1/private/overview/pipeline*` endpoints for listing and
 * counting runs, plus `/api/v1/private/runs/:runId/traces` for trace access
 * and `/api/v1/private/pipeline-schedule` for cron scheduling.
 */
import type { HttpClient } from "./client.js";
import type {
  DeleteTracesResult,
  OverviewSummary,
  PipelineCountByStatus,
  PipelineListParams,
  PipelineRun,
  PipelineSchedule,
  PipelineScheduleUpsertBody,
  Trace,
  TracePage,
  TraceStats,
} from "./platform-types.js";

// ── Embeddable lifecycle types ────────────────────────────────────────────────
// These cover the `/api/v1/private/pipelines/*` surface that lets you build,
// run, and read agent pipelines end to end from your own app.

/** Full pipeline configuration (agents, prompts, models, tools, wiring). */
export type PipelineConfig = Record<string, unknown> & { name?: string; project?: string };

/** Body for `create()`. `name` + `project` are required; the rest is the
 *  pipeline config (agents, prompts, models, wiring). */
export interface PipelineCreateBody {
  name: string;
  project: string;
  description?: string;
  [key: string]: unknown;
}

/** Options for `run()`. All optional — omit for a plain cloud-spawn run. */
export interface PipelineRunOptions {
  /** Project the pipeline belongs to (narrows tenant scope). */
  project?: string;
  /** Where the run executes. Defaults to the pipeline's own setting. */
  executionTarget?: "local-runner" | "cloud-spawn";
  /** Optional studio URL for run-event callbacks. */
  studio_url?: string;
  /** Per-run env var overrides layered over the caller's stored credentials. */
  env_overrides?: Record<string, string>;
}

/** Accepted-run envelope returned by `run()`. */
export interface PipelineRunAccepted {
  run_id: string;
  queued: boolean;
}

/** Run status + best-effort cost returned by `runStatus()`. */
export interface PipelineRunStatus {
  runId: string;
  status: string;
  createdAt: string;
  executionTarget: "local-runner" | "cloud-spawn";
  /** Cost breakdown from the in-process tracker; null for subprocess runs. */
  cost: unknown | null;
}

/** Body for `instantiateTemplate()`. */
export interface TemplateInstantiateBody {
  name: string;
  project: string;
  /** Config keys to override on top of the template payload. */
  overrides?: Record<string, unknown>;
}

const enc = encodeURIComponent;

export class PipelinesAPI {
  constructor(private readonly http: HttpClient) {}

  // ── Embeddable lifecycle: build → run → read ────────────────────────────────
  // Full CRUD + run + outputs over `/api/v1/private/pipelines/*`. Every method
  // is tier-gated server-side (Forge+ to create/run) and tenant-scoped to the
  // caller's projects. Prompt/model edits per agent go through `update()` with
  // the full config body.

  /** List pipelines visible to the caller. */
  async listPipelines(): Promise<{ pipelines: PipelineConfig[] }> {
    return this.http.get<{ pipelines: PipelineConfig[] }>("/api/v1/private/pipelines");
  }

  /**
   * Create a pipeline. `name` + `project` are required; include the agent
   * config (agents, prompts, models, wiring) in the same body.
   *
   * @example
   * ```ts
   * await melaya.agents.pipelines.create({
   *   name: "daily-digest", project: "acme",
   *   agents: [{ id: "researcher", model: "claude-sonnet-4-6", prompt: "..." }],
   * });
   * ```
   */
  async create(body: PipelineCreateBody): Promise<PipelineConfig> {
    return this.http.post<PipelineConfig>("/api/v1/private/pipelines", body);
  }

  /** Get one pipeline's full config. */
  async get(name: string, project?: string): Promise<PipelineConfig> {
    return this.http.get<PipelineConfig>(
      `/api/v1/private/pipelines/${enc(name)}`,
      project ? { project } : undefined,
    );
  }

  /**
   * Update a pipeline's config — the path for editing per-agent prompts or
   * swapping a model on one or all agents. Pass the full config plus `project`.
   *
   * @example
   * ```ts
   * const cfg = await melaya.agents.pipelines.get("daily-digest", "acme");
   * cfg.agents = cfg.agents.map(a => ({ ...a, model: "claude-opus-4-8" })); // all agents
   * await melaya.agents.pipelines.update("daily-digest", cfg, "acme");
   * ```
   */
  async update(name: string, config: PipelineConfig, project: string): Promise<PipelineConfig> {
    return this.http.put<PipelineConfig>(`/api/v1/private/pipelines/${enc(name)}`, {
      config,
      project,
    });
  }

  /** Delete a pipeline. */
  async remove(name: string, project: string): Promise<{ ok?: boolean } & Record<string, unknown>> {
    return this.http.delete(`/api/v1/private/pipelines/${enc(name)}`, { project });
  }

  /**
   * Run a pipeline. Returns immediately with a `run_id`; subscribe to progress
   * via `melaya.platform.events.onRunUpdate(run_id, ...)` or poll `runStatus()`.
   *
   * @example
   * ```ts
   * const { run_id } = await melaya.agents.pipelines.run("daily-digest", { project: "acme" });
   * melaya.platform.events.onRunUpdate(run_id, (e) => console.log(e.event_type, e.status));
   * ```
   */
  async run(name: string, opts?: PipelineRunOptions): Promise<PipelineRunAccepted> {
    return this.http.post<PipelineRunAccepted>(
      `/api/v1/private/pipelines/${enc(name)}/run`,
      opts ?? {},
    );
  }

  /** List run IDs for a pipeline (filtered to the caller's tier retention window). */
  async runIds(name: string): Promise<{ run_ids: string[] } & Record<string, unknown>> {
    return this.http.get(`/api/v1/private/pipelines/${enc(name)}/runs`);
  }

  /** Get status + best-effort cost for one run. */
  async runStatus(name: string, runId: string): Promise<PipelineRunStatus> {
    return this.http.get<PipelineRunStatus>(
      `/api/v1/private/pipelines/${enc(name)}/runs/${enc(runId)}`,
    );
  }

  /** Cancel / kill an in-flight run. */
  async cancelRun(name: string, runId: string): Promise<{ ok?: boolean } & Record<string, unknown>> {
    return this.http.delete(`/api/v1/private/pipelines/${enc(name)}/runs/${enc(runId)}`);
  }

  /** List the artifacts a pipeline has produced. */
  async outputs(name: string): Promise<unknown> {
    return this.http.get(`/api/v1/private/pipelines/${enc(name)}/outputs`);
  }

  /**
   * Read one output artifact by relative path. Text/JSON artifacts are parsed;
   * pass `{ download: true }` to force a download disposition on the wire.
   */
  async output(name: string, path: string, opts?: { download?: boolean }): Promise<unknown> {
    const rel = path.split("/").map(enc).join("/");
    return this.http.get(
      `/api/v1/private/pipelines/${enc(name)}/outputs/${rel}`,
      opts?.download ? { download: "1" } : undefined,
    );
  }

  /** Read-only codegen preview for a config (no persistence, no tier gate). */
  async previewCode(config: PipelineConfig): Promise<unknown> {
    return this.http.post(`/api/v1/private/pipelines/preview-code`, config);
  }

  /** The builder tool registry available to pipeline agents. */
  async tools(): Promise<unknown> {
    return this.http.get(`/api/v1/private/pipelines/tools`);
  }

  /** The builder sub-agent / crew registry. */
  async subagents(): Promise<unknown> {
    return this.http.get(`/api/v1/private/pipelines/subagents`);
  }

  /**
   * Live platform catalog counts (agentic tools, subagents, by category). Public.
   * Moved here from `MarketAPI` — use `melaya.agents.pipelines.catalogCounts()`.
   */
  async catalogCounts(): Promise<{ tools: number; subagents: number; byCategory?: unknown }> {
    return this.http.get<{ tools: number; subagents: number; byCategory?: unknown }>("/api/v1/public/catalog-counts");
  }

  /**
   * Dashboard usage summary (pipeline count, RAG usage, plan limits) for the sidebar/overview.
   * Moved here from `AccountAPI` — use `melaya.agents.pipelines.usageSummary()`.
   */
  async usageSummary(): Promise<Record<string, unknown>> {
    return this.http.get<Record<string, unknown>>("/api/v1/private/overview/usage");
  }

  /**
   * Instantiate a pipeline from a template you can access. Merges the template
   * payload with your `overrides`, then creates it under `name` + `project`.
   */
  async instantiateTemplate(
    templateId: string,
    body: TemplateInstantiateBody,
  ): Promise<{ pipeline: PipelineConfig }> {
    return this.http.post<{ pipeline: PipelineConfig }>(
      `/api/v1/private/templates/${enc(templateId)}/instantiate`,
      body,
    );
  }

  /**
   * Build a pipeline from a natural-language brief (synchronous — returns the
   * generated config). For streamed progress, POST to
   * `/api/v1/private/ai/build-pipeline` and read the SSE stream directly.
   */
  async buildWithAI(brief: Record<string, unknown>): Promise<unknown> {
    return this.http.post(`/api/v1/private/ai/build-pipeline/sync`, brief);
  }

  // ── Overview ────────────────────────────────────────────────────────────────

  /** Dashboard overview: usage stats, active strategies, recent runs. */
  async overview(): Promise<OverviewSummary> {
    return this.http.get<OverviewSummary>("/api/v1/private/overview");
  }

  async modelPrices(): Promise<Record<string, unknown>> {
    return this.http.get("/api/v1/private/overview/model-prices");
  }

  async chartData(params: Record<string, string | number | boolean | undefined> = {}): Promise<Record<string, unknown>> {
    return this.http.get("/api/v1/private/overview/chart", params);
  }

  async costBreakdown(params: Record<string, string | number | boolean | undefined> = {}): Promise<Record<string, unknown>> {
    return this.http.get("/api/v1/private/overview/cost-breakdown", params);
  }

  /** Count of pipeline runs grouped by status. */
  async count(): Promise<PipelineCountByStatus> {
    return this.http.get<PipelineCountByStatus>("/api/v1/private/overview/pipeline-count");
  }

  /**
   * Paginated list of pipeline runs.
   *
   * @example
   * ```ts
   * const runs = await melaya.pipelines.list({ project: "my-project", limit: 20 });
   * ```
   */
  async list(params?: PipelineListParams): Promise<PipelineRun[]> {
    return this.http.get<PipelineRun[]>("/api/v1/private/overview/pipelines", {
      ...(params as Record<string, string | number | boolean | undefined | null>),
    });
  }

  /** Most recent pipeline runs for a dashboard widget. */
  async recent(): Promise<PipelineRun[]> {
    return this.http.get<PipelineRun[]>("/api/v1/private/overview/pipelines/recent");
  }

  async serverVersion(): Promise<Record<string, unknown>> {
    return this.http.get("/api/v1/version");
  }

  // ── Traces ──────────────────────────────────────────────────────────────────

  /**
   * List traces for a run.
   * Returns a paginated envelope: `{ data: { list, total, page, pageSize } }`.
   * Access results via `.data.list`.
   */
  async traces(runId: string, params?: { page?: number; pageSize?: number }): Promise<{ data: TracePage }> {
    return this.http.get<{ data: TracePage }>(
      `/api/v1/private/runs/${encodeURIComponent(runId)}/traces`,
      params as Record<string, string | number | undefined>,
    );
  }

  /** Get a single trace by ID. */
  async trace(runId: string, traceId: string): Promise<Trace> {
    return this.http.get<Trace>(
      `/api/v1/private/runs/${encodeURIComponent(runId)}/traces/${encodeURIComponent(traceId)}`,
    );
  }

  /** Get statistics for a specific trace. */
  async traceStats(runId: string, traceId: string): Promise<TraceStats> {
    return this.http.get<TraceStats>(
      `/api/v1/private/runs/${encodeURIComponent(runId)}/traces/${encodeURIComponent(traceId)}/stats`,
    );
  }

  /**
   * Delete all traces (and their spans) for a run.
   * Issues DELETE /api/v1/private/runs/:runId/traces — no request body.
   */
  async deleteTraces(runId: string): Promise<DeleteTracesResult> {
    return this.http.delete<DeleteTracesResult>(
      `/api/v1/private/runs/${encodeURIComponent(runId)}/traces`,
    );
  }

  // ── Schedule ────────────────────────────────────────────────────────────────

  /** List all pipeline schedules accessible to the caller. */
  async listSchedules(): Promise<PipelineSchedule[]> {
    return this.http.get<PipelineSchedule[]>("/api/v1/private/pipeline-schedule");
  }

  /** Get the schedule for a specific pipeline. */
  async getSchedule(project: string, pipelineName: string): Promise<PipelineSchedule> {
    return this.http.get<PipelineSchedule>(
      `/api/v1/private/pipeline-schedule/${encodeURIComponent(project)}/${encodeURIComponent(pipelineName)}`,
    );
  }

  /**
   * Create or update a pipeline schedule (cron expression + optional config).
   *
   * @example
   * ```ts
   * await melaya.pipelines.upsertSchedule("my-project", "nightly-report", { cron: "0 2 * * *" });
   * ```
   */
  async upsertSchedule(
    project: string,
    pipelineName: string,
    body: PipelineScheduleUpsertBody,
  ): Promise<PipelineSchedule> {
    return this.http.put<PipelineSchedule>(
      `/api/v1/private/pipeline-schedule/${encodeURIComponent(project)}/${encodeURIComponent(pipelineName)}`,
      body,
    );
  }

  /** Pause a pipeline schedule. */
  async pauseSchedule(project: string, pipelineName: string): Promise<{ ok: boolean }> {
    return this.http.post<{ ok: boolean }>(
      `/api/v1/private/pipeline-schedule/${encodeURIComponent(project)}/${encodeURIComponent(pipelineName)}/pause`,
    );
  }

  /** Resume a paused pipeline schedule. */
  async resumeSchedule(project: string, pipelineName: string): Promise<{ ok: boolean }> {
    return this.http.post<{ ok: boolean }>(
      `/api/v1/private/pipeline-schedule/${encodeURIComponent(project)}/${encodeURIComponent(pipelineName)}/resume`,
    );
  }
}
